### ISSUE TITLE
[3.0] Update Emogrifier class for Php7.1 support and bugfixes

### ISSUE THREAD

**Issue by unfulvio** [woocommerce/woocommerce#13788]
## EXPLANATION OF THE ISSUE

I noticed that in WC 3.0 when sending an email that had  a `<style type="text/css>/* ...stuff... */ </style>` tag, this is parsed by the Emogrifier class. For some reason in WC 3.0 the same email using the same code, HTML and CSS had Emogrifier throwing a notice which I did not see in WC 2.6.14 or earlier

<pre><code>
[28-Mar-2017 04:22:10 UTC] PHP Warning:  DOMXPath::query(): Invalid expression in /library/woocommerce/includes/libraries/class-emogrifier.php on line 289
[28-Mar-2017 04:22:10 UTC] PHP Stack trace:
[28-Mar-2017 04:22:10 UTC] PHP   1. {main}() /wp-cron.php:0
[28-Mar-2017 04:22:10 UTC] PHP   2. do_action_ref_array() /wp-cron.php:117
[28-Mar-2017 04:22:10 UTC] PHP   3. WP_Hook->do_action() /wp-includes/plugin.php:515
[28-Mar-2017 04:22:10 UTC] PHP   4. WP_Hook->apply_filters() /wp-includes/class-wp-hook.php:323
[28-Mar-2017 04:22:10 UTC] PHP   5. WC_Emails::send_queued_transactional_email() /wp-includes/class-wp-hook.php:298
[28-Mar-2017 04:22:10 UTC] PHP   6. do_action_ref_array() /library/woocommerce/includes/class-wc-emails.php:134
[28-Mar-2017 04:22:10 UTC] PHP   7. WP_Hook->do_action() /wp-includes/plugin.php:515
[28-Mar-2017 04:22:10 UTC] PHP   8. WP_Hook->apply_filters() /wp-includes/class-wp-hook.php:323
[28-Mar-2017 04:22:10 UTC] PHP   9. WC_PIP_Email_Packing_List->trigger() /wp-includes/class-wp-hook.php:300
[28-Mar-2017 04:22:10 UTC] PHP  10. WC_Email->send() /woocommerce-pip/includes/emails/class-wc-pip-email-packing-list.php:211
[28-Mar-2017 04:22:10 UTC] PHP  11. WC_Email->style_inline() /library/woocommerce/includes/emails/class-wc-email.php:471
[28-Mar-2017 04:22:10 UTC] PHP  12. Emogrifier->emogrify() /library/woocommerce/includes/emails/class-wc-email.php:418
[28-Mar-2017 04:22:10 UTC] PHP  13. Emogrifier->process() /library/woocommerce/includes/libraries/class-emogrifier.php:214
[28-Mar-2017 04:22:10 UTC] PHP  14. DOMXPath->query() /library/woocommerce/includes/libraries/class-emogrifier.php:289
</pre></code>

I narrowed down this issue to my CSS rules having this type of case which has been covered by a most recent Emogrifier version:

https://github.com/jjriv/emogrifier/pull/308

but also see https://github.com/jjriv/emogrifier/pull/361

There are more kinds of valid CSS rules that generated bugs in Emogrifier, see their changelog covering releases after v1.0.0 which is the version bundled by WooCommerce

https://github.com/jjriv/emogrifier/releases

their latest version also supports PHP 7.1 --- updating this class in WooCommerce 3.0 solved the issue for me. I think if you give additional testing to it, it could be a good reason for WooCommerce to update as well to get the most recent version benefits and bugfixes.

## STEPS TO REPRODUCE THE ISSUE

Try having an email with a `<style>` block containing CSS rules that have been covered in bugfixes in Emogrfier after version 1.0.0

## SYSTEM STATUS REPORT

does not apply here, I'm running WC 3.0 and Print Invoices & Packing Lists, if I switch back to WC 2.6 this issue does not show up with no change in my email HTML or CSS. I tried commenting out parts until I narrowed down the email HTML to the aforementioned styles which trigger Emogrifier 1.0.0 bugs and consequent PHP notice

**Comment by unfulvio** [comment]
update: I realized that newer versions of Emogrifier contain PHP 5.3/5.4 code (mainly top level namespaces and PHP 5.4 notation for arrays). 

This doesn't change the fact the problem reported above persists though, it would be ideal to find a solution or allow plugins to address the issue with their own workaround.

**Comment by mikejolley** [comment]
I closed https://github.com/woocommerce/woocommerce/pull/13789 because of the non-5.2 compatible stuff. We could make a custom version to get around this but I'l like to leave it for future updates since we're in RC.

I'll also bring up internally the 5.2 discussion as I'd like to drop it asap.

**Comment by claudiosanches** [comment]
@unfulvio I'm using PHP 7.1 since has been released in my dev environment and I'm not getting this error, I receive all emails as HTML and text...

> Try having an email with a <style> block containing CSS rules that have been covered in bugfixes in Emogrfier after version 1.0.0

Not sure what this mean, can you give an example?
So maybe I can try fix without have to update.

**Comment by unfulvio** [comment]
oi @claudiosanches 

I bumped in this problem with Print Invoices & Packing Lists

1. Get WooCommerce 3.0
2. Get the latest public released version of PIP https://github.com/woocommerce/woocommerce-pip/releases/tag/3.2.0
3. Uncomment this if statement: https://github.com/woocommerce/woocommerce-pip/blob/master/templates/pip/styles.php#L96-L103
4. Process an order, then send an email with PIP 
5. You may notice a `PHP Warning:  DOMXPath::query(): Invalid expression in /library/woocommerce/includes/libraries/class-emogrifier.php on line 289` PHP warning notice while email is sent from WC Orders admin edit screen

so, for me it got triggered while having this CSS selector in a `<style>` block inside an email template:

`body > div.container:last-child:after `

which should be legit CSS, but from what I've understood Emogrifier could raise bugs with nested pseudo selectors like these - I have no idea why I didn't get this error in 2.6.14 but only WC 3.0

anyhow, there are indeed bugs covered since Emogrifier 1.0 and then added Php 7.1 support according tot he changelogs. So, I believe some plugin, some customization in the HTML/CSS emails, might produce errors in WC because of an unpatched Emogrifier class bundled in WC. Just check their changelogs.

In the case of PIP it looked like that excluding that rule did the trick as workaround (it wasn't so important), but thought of raising the issue here on WC as the problem is still there.

Thanks

**Comment by claudiosanches** [comment]
Don't sounds like an error on PHP 7.0 or 7.1, I think that just require improve the regex to support `:last-child:after`. I'll try fix it.

**Comment by unfulvio** [comment]
yeah it's unlikely the issue, but still in a future WC release perhaps you'll want to go through the bugs fixed by Emogrifier and see what could be backported

obrigado

**Comment by claudiosanches** [comment]
@unfulvio Have you tested the last version? I just updated here and using the code from PIP it's just logged an error in the `emogrifier-*.log` file instead of the `wp-content/debug.log` because they improve how handle exceptions. For the docs that I found in the code and on GitHub they don't plan to support `.container:last-child:after`.
In any case I recommend you guys to stop use this selector on PIP.
